lib/Arch/X86_C.cpp

```diff
@@ -0,0 +1,152 @@
+  // the callee is responbile for returning said sret argument in %eax and
+  // cleaning up the sret argument with a `ret 4`. This changes the
+  // return_stack_pointer offset because it will now be 4 bytes higher than we
+  // thought.
+  //
+  // However, even if there is sret on the second argument as well, we do not
+  // need to worry about this. For some reason the callee is only responsible
+  // for cleaning up the case where an sret argument is passed in as the first
+  // argument.
+  if (func.hasParamAttribute(0, llvm::Attribute::StructRet)) {
+    fdecl.return_stack_pointer_offset = 8;
+  } else {
+    fdecl.return_stack_pointer_offset = 4;
+  }
+
+  fdecl.return_stack_pointer = arch->RegisterByName("ESP");
+}
+
+std::vector<anvill::ValueDecl> X86_C::BindReturnValues(
+    const llvm::Function &function, bool &injected_sret) {
+  std::vector<anvill::ValueDecl> return_value_declarations;
+  anvill::ValueDecl value_declaration = {};
+  injected_sret = false;
+
+  // If there is an sret parameter then it is a special case. For the X86_C ABI,
+  // the sret parameters are guarenteed the be in %eax. In this case, we can
+  // assume the actual return value of the function will be the sret struct
+  // pointer.
+  for (auto arg = function.arg_begin(); arg != function.arg_end(); arg++) {
+    if (arg->hasStructRetAttr()) {
+      value_declaration.type = arg->getType();
+      value_declaration.reg = arch->RegisterByName("EAX");
+      return_value_declarations.push_back(value_declaration);
+      return return_value_declarations;
+    }
+  }
+
+  llvm::Type *ret_type = function.getReturnType();
+  value_declaration.type = ret_type;
+
+  switch (ret_type->getTypeID()) {
+    case llvm::Type::IntegerTyID:
+    case llvm::Type::PointerTyID: {
+      // Allocate EAX for an integer or pointer
+      value_declaration.reg = arch->RegisterByName("EAX");
+      break;
+    }
+    case llvm::Type::FloatTyID: {
+      // Allocate ST0 for a floating point value
+      value_declaration.reg = arch->RegisterByName("ST0");
+      break;
+    }
+    case llvm::Type::StructTyID: {
+      // Try to split the struct over the registers
+      std::vector<bool> allocated(return_register_constraints.size(), false);
+      auto struct_ptr = llvm::cast<llvm::StructType>(value_declaration.type);
+      auto mapping =
+          TryReturnThroughRegisters(*struct_ptr, return_register_constraints);
+      if (mapping) {
+        // There is a valid split over registers, so return the mapping
+        return *mapping;
+      } else {
+        // Struct splitting didn't work so do RVO. Assume that the pointer
+        // to the return value resides in EAX. In this case we have injected an
+        // sret into the first parameter so we need to take note of that.
+        injected_sret = true;
+        value_declaration.reg = arch->RegisterByName("EAX");
+      }
+      break;
+    }
+    default: {
+      LOG(ERROR) << "Encountered an unknown return type";
```

Comment:
Elsewhere you have `LOG(FATAL)`, perhaps use that here. Also, I recommend adding in `remill::LLVMThingToString(ret_type)` to make the logs more actionable.

Reply:
Fixed 028b2a5